### PR TITLE
ROX-19008: Add scanner config and use it

### PR DESCRIPTION
**Pull request by jvdm**
## Description

Adds `-conf` to scanner cmd, which reads and parse an YAML that has entries to configure multiple parts of scanner.

On deployment, `-conf` will receive a file from a config map.

## Checklist

- [ ] Investigated and inspected CI test results
- [X] Unit test and regression tests added

### N/A

- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

### Here I tell how I validated my change

Local `go run scanner/cmd/scanner/main.go -conf config.yaml` with the following config:

```yaml
matcher:
  enable: false
indexer:
  db_conn_string: postgresql:///jvdm?host=/var/run/postgresql
log_level: debug
mtls:
  certs_dir: /home/jvdm/tmp/scanner-tls
```

Played with different configs.

Also, UTs for the config validation.

Deployed images to an OCP cluster, and was able to run scanner and load config from config map.

### CODE REVIEW

File: scanner/config/config.go [hunk 1]
```diff
@@ -1,9 +1,143 @@
...
+	"io"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/rs/zerolog"
+	"gopkg.in/yaml.v3"
+)
+
+var (
```
Comment (daynewlee) [inline scanner/config/config.go:14]: Maybe we can have a function:
```
func NewDefaultConfiguration() Config {
    return Config{
        HTTPListenAddr: ":9443",
        GRPCListenAddr: ":8443",
        // ... other values ...
    }
}
```
And use it this way:
`cfg := NewDefaultConfiguration()`

That should make the default config immutable.

Comment (jvdm) [reply scanner/config/config.go:14]: I don't understand what you mean by "making the default config immutable". Can you please clarify?

File: scanner/config/config.go [hunk 2]
```diff
@@ -1,9 +1,143 @@
...
+	if err != nil {
+		return fmt.Errorf("mtls: %w", err)
+	}
+	return nil
+}
+
+// IndexerConfig provides Scanner Indexer configuration.
+type IndexerConfig struct {
+	// Database provides indexer's database configuration.
+	DBConnString string `yaml:"db_conn_string"`
```
Comment (RTann) [inline scanner/config/config.go:56]: I think Central hardcodes this (unless it is connecting to someone's BYODB). I think it may be worth asking why they chose to do that to see if we would want to consider that as well.

We could also opt to continue as-is and make it configurable, but then we could consider just using a hardcoded default (and then updating the default above to one which is more complete (has sslmode, password stuff, etc))

File: scanner/config/config.go [hunk 3]
```diff
@@ -1,9 +1,143 @@
...
 type Config struct {
-	// Mode specifies the mode in which Scanner will run.
-	//
-	// The options are: Indexer, Matcher, Combo.
-	Mode string `json:"mode"`
+	Indexer        IndexerConfig `yaml:"indexer"`
+	Matcher        MatcherConfig `yaml:"matcher"`
+	HTTPListenAddr string        `yaml:"http_listen_addr"`
+	GRPCListenAddr string        `yaml:"grpc_listen_addr"`
+	MTLS           MTLSConfig    `yaml:"mtls"`
```
Comment (RTann) [inline scanner/config/config.go:41]: up to you, but I think just TLS is sufficient. Don't think we need to make it clear that we use mTLS

File: scanner/config/config.go [hunk 4]
```diff
@@ -1,9 +1,143 @@
...
+	"time"
+
+	"github.com/rs/zerolog"
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	// DefaultConfiguration provides the default values for the scanner configuration.
+	DefaultConfiguration = Config{
+		HTTPListenAddr: ":9443",
```
Comment (RTann) [inline scanner/config/config.go:17]: Curious where did this come from? From what I can tell, metrics port tends to be 9090 or so (https://github.com/stackrox/stackrox/blob/4.1.3/pkg/env/metrics.go#L6)

File: scanner/config/config.go [hunk 5]
```diff
@@ -1,9 +1,143 @@
...
+
 // Config represents the Scanner configuration parameters.
 type Config struct {
-	// Mode specifies the mode in which Scanner will run.
-	//
-	// The options are: Indexer, Matcher, Combo.
-	Mode string `json:"mode"`
+	Indexer        IndexerConfig `yaml:"indexer"`
+	Matcher        MatcherConfig `yaml:"matcher"`
+	HTTPListenAddr string        `yaml:"http_listen_addr"`
```
Comment (RTann) [inline scanner/config/config.go:39]: Curious why you chose address instead of just an `int` port?

File: scanner/config/config.go [hunk 6]
```diff
@@ -1,9 +1,143 @@
...
+	GRPCListenAddr string        `yaml:"grpc_listen_addr"`
+	MTLS           MTLSConfig    `yaml:"mtls"`
+	LogLevel       LogLevel      `yaml:"log_level"`
+}
+
+func (c *Config) validate() error {
```
Comment (RTann) [inline scanner/config/config.go:45]: if we continue with HTTP/gRPC address, can we validate those, too?

File: scanner/config/config.go [hunk 7]
```diff
@@ -1,9 +1,143 @@
...
+}
+
+// IndexerConfig provides Scanner Indexer configuration.
+type IndexerConfig struct {
+	// Database provides indexer's database configuration.
+	DBConnString string `yaml:"db_conn_string"`
+	// Enable if false disables the Indexer service.
+	Enable bool `yaml:"enable"`
+	// GetLayerTimeout timeout duration of GET requests for layers
+	GetLayerTimeout Duration `yaml:"get_layer_timeout"`
```
Comment (RTann) [inline scanner/config/config.go:60]: Do we want to set any default for this? For just a GET, maybe 30 seconds or so is a decent timeout? I'm thinking if we offer this option, we should set a default, as I can only imagine customers setting this themselves to make it longer (or infinite)

File: scanner/indexer/indexer.go [hunk 8]
```diff
@@ -79,7 +80,8 @@ func NewIndexer(ctx context.Context) (Indexer, error) {
 	}
 
 	return &indexerImpl{
-		indexer: indexer,
+		indexer:         indexer,
+		getLayerTimeout: time.Duration(cfg.GetLayerTimeout),
```
Comment (RTann) [inline scanner/indexer/indexer.go:84]: Will this var be used in another PR?

File: scanner/config/config.go (hunk 1 above)
Comment (RTann) [reply scanner/config/config.go:14]: I think it's fine the way it is

Comment (jvdm) [reply scanner/config/config.go:14]: Let's hear from Yi what his concern is about then we will decide if its fine or not.

File: scanner/config/config.go (hunk 2 above)
Comment (jvdm) [reply scanner/config/config.go:63]: > I think it may be worth asking why they chose to do that to see if we would want to consider that as well.

Hard coding is a subset of making it configurable with a default.

> We could also opt to continue as-is and make it configurable, but then we could consider just using a hardcoded default

Isn't t that what the patch is doing already?

> and then updating the default above to one which is more complete (has sslmode, password stuff, etc

The defaults were defined to run Scanner V4 in development/test mode, because that's where defaults are likely to make sense. In deployment, values should **always** be provided by the specs.

File: scanner/config/config.go (hunk 3 above)
Comment (jvdm) [reply scanner/config/config.go:42]: There are files in that directory related to mTLS, so sing MTLS here makes it more obvious what everything inside the directory is.

File: scanner/config/config.go (hunk 4 above)
Comment (jvdm) [reply scanner/config/config.go]: A different HTTP server serves Prometheus metrics. The port is configurable through an env called `ROX_METRICS_PORT`.

File: scanner/config/config.go (hunk 5 above)
Comment (jvdm) [reply scanner/config/config.go:40]: Because with sockets, you need to specify the binding address and port when defining endpoints.

File: scanner/config/config.go (hunk 1 above)
Comment (daynewlee) [reply scanner/config/config.go:15]: Discussed with Victor, we can make the variable a package-level var. And only expose copies of the default, so modifications to it will not change the actual default variable.
We would still be able to change the default inadvertently inside the config package, but at least it will be contained so less risk.

**Review by daynewlee — APPROVED** [review]
LGTM
